IC_TEENSY3_1 board config wakes from sleep successfully

djfiander opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description
The PJRC_TEENSY_3_1 board configuration will wake the computer when it's asleep, but then the keyboard wlll
lock up until it's unplugged and reconnected. Using the IC_TEENSY_3_1 board config fixes the problem. The computer will wake up, and the keyboard will continue to function.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [x ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

fauxpark commented:
This board is for the Input:Club keyboards, and has some Kiibohd-specific code. It should not be used for all Teensy boards.

sigprof commented:
Looks like another problem caused by the “fix” in #10088 — that change needed to be effectively reverted for Input Club boards, and that was done in #12870, but the empty replacement for `restart_usb_driver()` was placed in a board file which apparently also has some code specific to the Input Club boards.

You can try removing `BOARD = IC_TEENSY_3_1` and then adding this to `<keyboard>.c`:
```c
void restart_usb_driver(USBDriver *usbp) {
    // Do nothing. Restarting the USB driver on these boards breaks it.
}
```
But this will fix the issue only for one specific keyboard.

Ideally the original problem that prompted the #10088 hotfix should be found and fixed properly. Apparently the original issue happens mostly on STM32F4x1 chips with USB-OTG, and there is an attempt to do a somewhat dirty fix just for those chips in #14988, but there does not seem to be any movement there for some time.

djfiander commented:
I would really like it if both STM32 and NXP (teensy) boards both worked well, but digging that far into the code is way beyond me at this point. I'm just using the onekey as a test harness and framework for building my own keyboard(s).

If you've got suggestions for where to look to try to help develop the "real solution", just let me know. I'm not afraid of reading microcontroller manuals.

djfiander commented:
So, I removed the "BOARD=", and copied the "restart_usb_driver" into my local board. It works properly on my FreeBSD box, but not on my Mac. The Mac recognizes that the device has connected, but typing doesn't do anything.

> Looks like another problem caused by the “fix” in #10088 — that change needed to be effectively reverted for Input Club boards, and that was done in #12870, but the empty replacement for `restart_usb_driver()` was placed in a board file which apparently also has some code specific to the Input Club boards.
> 
> You can try removing `BOARD = IC_TEENSY_3_1` and then adding this to `<keyboard>.c`:
> 
> ```c
> void restart_usb_driver(USBDriver *usbp) {
>     // Do nothing. Restarting the USB driver on these boards breaks it.
> }
> ```
> 
> But this will fix the issue only for one specific keyboard.
> 
> Ideally the original problem that prompted the #10088 hotfix should be found and fixed properly. Apparently the original issue happens mostly on STM32F4x1 chips with USB-OTG, and there is an attempt to do a somewhat dirty fix just for those chips in #14988, but there does not seem to be any movement there for some time.

sigprof commented:
> So, I removed the "BOARD=", and copied the "restart_usb_driver" into my local board. It works properly on my FreeBSD box, but not on my Mac. The Mac recognizes that the device has connected, but typing doesn't do anything.

Is this behavior repeatable?  Does the problem with Mac disappear when you are using `BOARD = IC_TEENSY_3_1` instead of just overriding `restart_usb_driver()`?

You can try adding the other piece of code from `IC_TEENSY_3_1` to your `<keyboard>.c` file (before the empty definition of `restart_usb_driver()` that you added), then do the same testing as before:

```c
#define WDOG_TMROUTL *(volatile uint16_t *)0x40052012

void early_hardware_init_pre(void) {
    // This is a dirty hack and should only be used as a temporary fix until this
    // is upstreamed.
    while (WDOG_TMROUTL < 2)
        ;  // Must wait for WDOG timer if already running, before jumping
}
```

(There are no other differences between `PJRC_TEENSY_3_1` and `IC_TEENSY_3_1` — the initial GPIO setup for these board definitions is exactly the same.)

djfiander commented:
Bah! Ignore everything I said about it not working on my Mac. The problem was with the usb hub I use with my mac. So, moving JUST the restart_usb_driver() code into the onekey/teensy_32 works everywhere.

fauxpark commented on keyboards/handwired/onekey/teensy_32/config.h:
```diff
@@ -16,7 +16,6 @@
 #pragma once
 
-// TODO: including this causes "error: expected identifier before '(' token" errors
 //#include "config_common.h"
```
```suggestion
// TODO: including this causes "error: expected identifier before '(' token" errors
//#include "config_common.h"
```
